### ISSUE TITLE
[QST] Is there a GPU-accelerated version of scipy's filtfilt routine available?

### ISSUE THREAD

**Issue by aaronpearlman** [rapidsai/cusignal#171]
Hello,

I was wondering if there is a GPU-accelerated version of scipy's `filtfilt` routine available in cusignal. If not, are there any plans to implement this?

I am looking for a way to speed-up this routine. If this is not already implemented or planned for the future, could the developers offer some advice for how I might be able to write a faster implementation?

Thanks,
Aaron

**Comment by awthomp** [comment]
Hi @aaronpearlman - we have not yet implemented filtfilt, as the current SciPy implementation uses lfilter. We did some work on this in the past, but struggled with parallelization of the DF-II structure. If there is an FFT method, please let us know!

We are also tracking some other work on parallel IIR filters and welcome your contributions!

I’m on my phone and will be in and out over the next two weeks, but I’ll be sure to update this question with some links. 

CC @mnicely

**Comment by aaronpearlman** [comment]
Would you be willing to share those links that you mentioned on developing parallel IIR filters, work toward parallelizing the DF-II structure, etc...? I would be willing to contribute, if I am able.

Thanks,
Aaron

**Comment by awthomp** [comment]
Hey @aaronpearlman -- you bet. Here was @mnicely's initial work with getting `lfilter` to run on the GPU: https://github.com/rapidsai/cusignal/pull/78. We removed it from the codebase due to poor performance. 

There's also some more discussion on IIR filter parallelization in the cupy project here: https://github.com/cupy/cupy/pull/3645#issuecomment-663351896

CC: @coderforlife

Reopening the question too :).

**Comment by coderforlife** [comment]
I have written a version of `filtfilt` using cupy and my `lfilter` that is semi-parallelized (about half of it is completely parallelized but the other half is only parallelized over individual signals and not samples). For `lfilter` I found that if there is a large number of samples you need several signals to parallelize over to make it competitive with scipy (in the specific example, it needed 64 signals with 16384 samples to be competitive). I have additional ideas on improving the speed but haven't had a chance to implement them yet.

**Comment by aaronpearlman** [comment]
Hey @coderforlife, this sounds like really nice work! Are your versions of `filtfilt` or `lfilter` in a public repository somewhere that I can fork?

I'd also be interesting in trying out those speed improvements as well, once you have a chance to implement them!

**Comment by coderforlife** [comment]
I just posted it to a gist: https://gist.github.com/coderforlife/17bf471a519a7d092e82f95f10e8f6e4

It isn't well-organized at the moment and also it requires https://github.com/coderforlife/cupy/tree/signaltools or you have to wait until https://github.com/cupy/cupy/pull/3748 is merged into cupy/master.

**Comment by AashinShazar** [comment]
Sorry to comment on a dated post but has `lfilter` functionality been implemented yet in cusignal? I noticed some of the issues regarding that question have been closed out.

**Comment by awthomp** [comment]
Hi @AashinShazar -- thanks for your question! We haven't implemented `lfilter` in cuSignal. Well, in reality we *did* implement it but depreciated our version as the real-life performance wasn't as good as we anticipated. Unfortunately, the DF-II technique has a data dependency that makes an algorithmically parallel version challenging. We do support FIR based filtering, and I strongly recommend using `firfilter` or `fftconvolve` as an alternative to `lfilter`.

**Comment by coderforlife** [comment]
I am back working on this. The previous implementation had the following features:

- All computations involving `b` were parallelized (either with a custom kernel or with a convolution)
- All independent signals were parallelized, but each signal serially performed their computations across `a`/`y`
- `zf`/`zi` usage wasn't completely finished

I am currently working on a [prefix-scan](https://en.wikipedia.org/wiki/Prefix_sum) based implementation like how cumsum/cumprod are computed on a GPU in parallel. For `lfilter` this is insanely more tricky then those, so I am starting with the naive approach and seeing if it shows any potential what-so-ever. I currently have a CPU-based program that is parallelized over `a` that works when `a` is between 2 and 4 elements long (1 element long is just a convolution and that is already fully parallelized) with a single signal. It shows some promise so far, but I am not fully convinced yet. Next step is to move it to the GPU.

**Comment by awthomp** [comment]
Thanks for the update, @coderforlife. We're noodling around with some ideas as well and will certainly share best practices and our work as it takes form.
